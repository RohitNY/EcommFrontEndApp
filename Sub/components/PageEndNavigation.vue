<template>
  <div
    :class="[cssTheme, $style.pEn]"
    :data-ui-api="ui"
    :data-ui-rendered="renderUi"
  >
    <!-- New implementation: Ui delivers a button collection Array -->
    <EcomButtonUiModule v-if="buttonCollection" :buttons="buttonCollection" />
    <!--  -->
    <span v-if="buttonText1 && shouldBtn1Post &&!buttonCollection" ref="button1">
      <EcomHttpPost @on-submit="buttonClicked('buttonClicked1')">
        <EcomButton
          :disabled="buttonClicked1"
          animate="flipInX faster delay-300ms"
          :class="[$style.btn, 'u-display-hidden']"
        >
          <span v-if="processing && buttonClicked1" :class="$style.processing">
            <EcomSpinner fg-color="#fafafa" kind="bars" />
          </span>
          <span v-else>
            <span ref="buttonText1" v-text="buttonText1" />
            <EcomArrow
              :class="$style.arrow"
              direction="right"
              size="lg"
              :position="arrowPosition"
            />
          </span>
        </EcomButton>
      </EcomHttpPost>
    </span>
    <span v-else-if="!buttonCollection" ref="button1">
      <EcomButton
        animate="flipInX faster delay-300ms"
        :class="[$style.btn, 'u-display-hidden']"
        :disabled="buttonClicked1"
        @click="advancePathExperience('buttonClicked1')"
      >
        <span v-if="processing && buttonClicked1" :class="$style.processing">
          <EcomSpinner fg-color="#fafafa" kind="bars" />
        </span>
        <span v-else>
          <span ref="buttonText1" v-text="buttonText1" />
          <EcomArrow
            :class="$style.arrow"
            direction="right"
            size="lg"
            :position="arrowPosition"
          />
        </span>
      </EcomButton>
    </span>
    <!--  -->
    <span v-if="buttonText2 && shouldBtn2Post &&!buttonCollection" ref="button2">
      <EcomHttpPost @on-submit="buttonClicked('buttonClicked2')">
        <EcomButton
          :disabled="buttonClicked2"
          animate="flipInX faster delay-300ms"
          :class="[$style.btn, 'u-display-hidden']"
        >
          <span v-if="processing && buttonClicked2" :class="$style.processing">
            <EcomSpinner fg-color="#fafafa" kind="bars" />
          </span>
          <span v-else>
            <span ref="buttonText2" v-text="buttonText2" />
            <EcomArrow
              :class="$style.arrow"
              direction="right"
              size="lg"
              :position="arrowPosition"
            />
          </span>
        </EcomButton>
      </EcomHttpPost>
    </span>
    <span v-else-if="buttonText2 &&!buttonCollection" ref="button2">
      <EcomButton
        animate="flipInX faster delay-300ms"
        :class="[$style.btn, 'u-display-hidden']"
        :disabled="buttonClicked2"
        @click="advancePathExperience('buttonClicked2')"
      >
        <span v-if="processing && buttonClicked2" :class="$style.processing">
          <EcomSpinner fg-color="#fafafa" kind="bars" />
        </span>
        <span v-else>
          <span ref="buttonText2" v-text="buttonText2" />
          <EcomArrow
            :class="$style.arrow"
            direction="right"
            size="lg"
            :position="arrowPosition"
          />
        </span>
      </EcomButton>
    </span>
    <!--  -->
    <p :class="$style.legal" :style="ensureLegibleTextColor">
      NO PURCHASE OR FEE NECESSARY TO ENTER.
      <br>A PURCHASE WON’T IMPROVE AN INDIVIDUAL’S CHANCE OF WINNING.
    </p>
  </div>
</template>

<script>
import get from 'lodash/get'
import EcomButtonUiModule from '../components/UiComponents/PageEndNavigation/ButtonUiModule.vue'

export default {
  name: 'EcomPageEndNavigation',
  components: {
    EcomButtonUiModule
  },
  props: {
    ui: { type: [String, Array], default: 'Ui1' },
    theme: { type: String, default: 'standard' }
  },
  data () {
    return {
      renderUi: null,
      buttonCollection: null,
      buttonText1: null,
      buttonText2: null,
      buttonClicked1: false,
      buttonClicked2: false,
      processing: false
    }
  },
  computed: {
    cssTheme () {
      const base = 'ecom-page-navigation'

      const cssClass = this.theme
        ? base + '--' + this.theme
        : base + '--standard'

      return cssClass
    },
    arrowPosition () {
      if (this.theme !== 'standard') {
        return 'borderRight'
      }
      return 'inline'
    },

    ensureLegibleTextColor () {
      const getter = this.$store.getters['app/getPagePathExperienceBySlug']
      const xp = getter(this.$route.query.tn)
      const color = this.$hopX.invertColor(xp?.Ui?.Colors?.PageBackground)
      return { color }
    },
    isCustomerOnFinalTab () {
      const path = this.$store.getters['app/getNextPathExperience']
      return typeof path.Slug === 'undefined'
    },
    /* ui rules */
    shouldBtn1Post () {
      return this.isCustomerOnFinalTab
    },
    shouldBtn2Post () {
      return this.renderUi === 'Ui3'
    },
    getSpectrumConfig () {
      return (
        get(window, 'Spectrum.UiComponentData.PageNavigation.' + this.ui) || {}
      )
    }
  },
  beforeMount () {
    if (Array.isArray(this.ui)) {
      // updated implementation with array of button components - Sept 2021
      // assume array has length < 2 and work with first two elements
      const buttonUiStrings = this.ui
      this.buttonCollection = buttonUiStrings.map((string) => {
        return window.Spectrum.UiComponentData.PageNavigation[string]
      })
    } else {
      this.renderUi = this.ui

      const { Labels } = this.getSpectrumConfig

      if (Labels) {
        this.buttonText1 = Labels[0]
        this.buttonText2 = Labels[1]
      }
    }
  },
  methods: {
    buttonClicked (buttonClicked) {
      this.processing = true
      this[buttonClicked] = true
    },
    async advancePathExperience (buttonClicked) {
      this.buttonClicked(buttonClicked)

      /* this sleep is to allow time for the btn ripple animation to play */
      await this.$hopX.sleep(350)
      this.$store.dispatch('app/continuePathExperience')
    }
  }
}
</script>

<style lang="scss" module>
.pEn {
  form {
    text-align: center;
    margin-bottom: 3rem;
    button {
      max-width: 27rem;
    }
  }
}

.btn {
  margin: 4rem auto 2rem;
  display: flex;
  justify-content: center;
  align-items: center;
  .ecomButtonStyle {
    background-color: var(--page-end-button-bg);
    font-family: var(--app-font-bold);
    max-width: calc(100% - 2rem);
    letter-spacing: -0.125rem;
    border-radius: var(--page-end-border-radius);
    color: #ffffff;
    box-shadow: 0px 3px 6px rgba(0, 0, 0, 0.161);
    z-index: 1;
    @include sm {
      font-size: 2rem;
      line-height: 2.5rem;
      min-width: 30rem;
      padding: 0 4rem;
      min-height: var(--page-end-min-height-sm);
    }
    @include md {
      font-size: 3.5rem;
      line-height: 4rem;
      padding: 0 2rem;
      min-width: 40rem;
      min-height: var(--page-end-min-height-md);
    }
    @include lg {
      font-size: 4rem;
      line-height: 4.5rem;
    }
  }
  .arrow > span {
    border-width: 0 0.6rem 0.6rem 0;
    padding: 1.2rem;
    @include sm {
      display: none;
    }
  }

  :disabled {
    transition: all 200ms cubic-bezier(0.075, 0.82, 0.165, 1);
    background-color: rgb(255, 220, 167);
    cursor: default;
  }
}

.processing {
  height: 5rem;
  display: block;
  margin-top: -8rem;
}

.legal {
  color: #000000; /* default */
  font-family: var(--app-font-bold);
  text-align: center;
  font-size: 1.222rem;
  line-height: 1.555rem;
  padding: 1rem 0;
}
</style>
