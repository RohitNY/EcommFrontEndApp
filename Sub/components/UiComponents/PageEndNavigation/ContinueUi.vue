<template>
  <span>
    <EcomButton
      animate="flipInX faster delay-300ms"
      :class="[$style.btn, 'u-display-hidden']"
      :disabled="clicked"
      @click="advancePathExperience('clicked')"
    >
      <span v-if="processing && clicked" :class="$style.processing">
        <EcomSpinner fg-color="#fafafa" kind="bars" />
      </span>
      <span v-else>
        <span v-text="buttonText" />
        <EcomArrow
          :class="$style.arrow"
          direction="right"
          size="lg"
          :position="arrowPosition"
        />
      </span>
    </EcomButton>
  </span>
</template>

<script>
export default {
  name: 'ContinueUi',
  props: {
    data: Object
  },
  data () {
    return {
      clicked: false,
      processing: false
    }
  },
  computed: {
    buttonText () {
      return this.data.Labels[0]
    },
    arrowPosition () {
      if (this.theme !== 'standard') {
        return 'borderRight'
      }
      return 'inline'
    }
  },
  methods: {
    buttonClicked (buttonClicked) {
      this.processing = true
      this[buttonClicked] = true
    },
    async advancePathExperience (buttonClicked) {
      this.buttonClicked(buttonClicked)

      /* this sleep is to allow time for the btn ripple animation to play */
      await this.$hopX.sleep(350)
      this.$store.dispatch('app/continuePathExperience')
    }
  }
}
</script>

<style lang="scss" module>
.btn {
  margin: 4rem auto 2rem;
  display: flex;
  justify-content: center;
  align-items: center;
  .ecomButtonStyle {
    background-color: var(--page-end-button-bg);
    font-family: var(--app-font-bold);
    max-width: calc(100% - 2rem);
    letter-spacing: -0.125rem;
    border-radius: var(--page-end-border-radius);
    color: #ffffff;
    box-shadow: 0px 3px 6px rgba(0, 0, 0, 0.161);
    z-index: 1;
    @include sm {
      font-size: 2rem;
      line-height: 2.5rem;
      min-width: 30rem;
      padding: 0 4rem;
      min-height: var(--page-end-min-height-sm);
    }
    @include md {
      font-size: 3.5rem;
      line-height: 4rem;
      padding: 0 2rem;
      min-width: 40rem;
      min-height: var(--page-end-min-height-md);
    }
    @include lg {
      font-size: 4rem;
      line-height: 4.5rem;
    }
  }
  .arrow > span {
    border-width: 0 0.6rem 0.6rem 0;
    padding: 1.2rem;
    @include sm {
      display: none;
    }
  }

  :disabled {
    transition: all 200ms cubic-bezier(0.075, 0.82, 0.165, 1);
    background-color: rgb(255, 220, 167);
    cursor: default;
  }
}

.processing {
  height: 5rem;
  display: block;
  margin-top: -8rem;
}
</style>
