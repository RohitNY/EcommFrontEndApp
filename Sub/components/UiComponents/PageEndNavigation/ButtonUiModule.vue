<script>
import ContinueUi from './ContinueUi.vue'
import SubmitUi from './SubmitUi.vue'

export default {
  name: 'EcomButtonUiModule',
  functional: true,
  props: {
    buttons: Array
  },
  render (h, ctx) {
    return (ctx.props.buttons.map((button) => {
      const { Component, ...data } = button
      const VueComponent = Component === 'SubmitUi' ? SubmitUi : ContinueUi
      return h(VueComponent, {
        props: {
          data
        }
      })
    }))
  }
}
</script>
