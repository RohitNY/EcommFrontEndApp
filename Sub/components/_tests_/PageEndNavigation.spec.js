import { mount, createLocalVue } from '@vue/test-utils'
import get from 'lodash/get'
import PageEndNavigation from '~/components/PageEndNavigation.vue'

/*
  Spectrum object reference
  ##################################################################################################
  {
    Ui1: {
      Component: 'ContinueUi1',
      Labels: [
        'CONTINUE'
      ]
    },
    Ui2: {
      Component: 'ContinueUi2',
      Labels: [
        'CONTINUE SHOPPING',
        'ENTER NOW' || CHECKOUT NOW (if cartCount > 0) => POST
      ]
    },
    Ui3: {
      Component: 'ContinueUi3',
      Labels: [
        'CONTINUE' => POST
      ]
    }
  }
*/

const localVue = createLocalVue()
localVue.directive('ui', {})
localVue.directive('ripple', {})

const getSpectrumValue = propsValue => (n) => {
  return get(
    global,
    `Spectrum.UiComponentData.PageNavigation.${propsValue}.${n}`
  )
}

const componentFactory = (overrides = {}, overrideMocks = {}) => {
  return Object.assign(
    {
      propsData: {
        ui: null
      },
      localVue,
      stubs: ['EcomArrow'],
      mocks: {
        $hopX: {
          sleep: () => {},
          invertColor: () => {},
          getParentComponent: () => {}
        },
        $route: {
          query: {
            tn: 'As-seen-on-tv'
          }
        },
        $store: {
          dispatch: () => {},
          getters: Object.assign(
            {
              'app/getNextPathExperience': {
                Slug: 'some-other-tab'
              },
              'app/getPagePathExperienceBySlug': () => ({
                Ui: {
                  Colors: {
                    PageBackground: '#ffffff'
                  }
                }
              })
            },
            overrideMocks.getters
          )
        }
      }
    },
    overrides
  )
}

// describe('PageEndNavigation', () => {
//   let wrapper

//   beforeEach(() => {
//     wrapper = mount(PageEndNavigation, componentFactory())
//   })

//   test('ensure legal disclaimer is rendered', () => {
//     const legal = wrapper.find('p.legal')

//     expect(legal).toBeTruthy()
//     expect(legal.text()).toContain('NO PURCHASE OR FEE NECESSARY TO ENTER.')
//     expect(legal.text()).toContain(
//       'A PURCHASE WON’T IMPROVE AN INDIVIDUAL’S CHANCE OF WINNING.'
//     )
//   })

//   // test('matches snapshot', () => {
//   //   expect(wrapper.html()).toMatchSnapshot()
//   // })
// })

// describe('[Ui1] PageEndNavigation with 1 button', () => {
//   let wrapper
//   const propsValue = 'Ui1'
//   const spectrum = getSpectrumValue(propsValue)

//   beforeEach(() => {
//     wrapper = mount(
//       PageEndNavigation,
//       componentFactory({
//         propsData: {
//           ui: propsValue
//         }
//       })
//     )
//   })

//   test('btn1 is rendered and NOT wrapped in HTML <form> tag', () => {
//     const btn1 = wrapper.findComponent({ ref: 'button1' })
//     expect(btn1.html()).not.toContain('<form')
//   })

//   // test('btn1 calls advancePathExperience method when clicked', async () => {
//   //   const btn1 = wrapper.findComponent({ ref: 'button1' }).find('button')
//   //   jest.spyOn(wrapper.vm, 'advancePathExperience')

//   //   await btn1.trigger('click')

//   //   expect(wrapper.vm.advancePathExperience).toHaveBeenCalled()
//   //   expect(btn1.html()).toContain('disabled')
//   // })

//   test('ensure Button label is correct based on Spectrum config for first btn', () => {
//     const btn1Text = wrapper.findComponent({ ref: 'buttonText1' })
//     expect(btn1Text).toBeTruthy()
//     expect(btn1Text.text()).toMatch(spectrum('Labels[0]'))
//   })

//   // test('matches snapshot', () => {
//   //   expect(wrapper.html()).toMatchSnapshot()
//   // })
// })

// describe('[Ui2] PageEndNavigation with 2 buttons', () => {
//   let wrapper
//   const propsValue = 'Ui2'
//   const spectrum = getSpectrumValue(propsValue)

//   beforeEach(() => {
//     wrapper = mount(
//       PageEndNavigation,
//       componentFactory({
//         propsData: {
//           ui: propsValue
//         }
//       })
//     )
//   })

//   test('ensure Button label is correct based on Spectrum config for first btn', () => {
//     const btn1Text = wrapper.findComponent({ ref: 'buttonText1' })
//     expect(btn1Text).toBeTruthy()
//     expect(btn1Text.text()).toMatch(spectrum('Labels[0]'))
//   })

//   test('ensure Button label is correct based on Spectrum config for second btn', () => {
//     const btn2Text = wrapper.findComponent({ ref: 'buttonText2' })
//     expect(btn2Text).toBeTruthy()
//     expect(btn2Text.text()).toMatch(spectrum('Labels[1]'))
//   })

//   test('btn1 is rendered and NOT wrapped in HTML <form> tag', () => {
//     const btn1 = wrapper.findComponent({ ref: 'button1' })
//     expect(btn1.html()).not.toContain('<form')
//   })

//   // test('btn2 is rendered wrapped in HTML <form> tag', () => {
//   //   const btn2 = wrapper.findComponent({ ref: 'button2' })
//   //   expect(btn2.html()).toContain('<form')
//   // })

//   // test('matches snapshot', () => {
//   //   expect(wrapper.html()).toMatchSnapshot()
//   // })
// })

// describe('[Ui3] PageEndNavigation with 1 button which will POST', () => {
//   let wrapper
//   const propsValue = 'Ui3'
//   const spectrum = getSpectrumValue(propsValue)

//   beforeEach(() => {
//     wrapper = mount(
//       PageEndNavigation,
//       componentFactory(
//         {
//           propsData: {
//             ui: propsValue
//           }
//         },
//         {
//           getters: {
//             'app/getNextPathExperience': {}
//           }
//         }
//       )
//     )
//   })

//   test('ensure Button label is correct based on Spectrum config for first btn', () => {
//     const btn1Text = wrapper.findComponent({ ref: 'buttonText1' })
//     expect(btn1Text).toBeTruthy()
//     expect(btn1Text.text()).toMatch(spectrum('Labels[0]'))
//   })

//   // test('btn1 is rendered wrapped in HTML <form> tag', () => {
//   //   const btn1 = wrapper.findComponent({ ref: 'button1' })
//   //   expect(btn1.html()).toContain('<form')
//   // })

//   // test('matches snapshot', () => {
//   //   expect(wrapper.html()).toMatchSnapshot()
//   // })
// })
