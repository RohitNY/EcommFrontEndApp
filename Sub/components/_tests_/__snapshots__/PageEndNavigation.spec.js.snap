// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Ui1] PageEndNavigation with 1 button matches snapshot 1`] = `
"<div data-ui-api=\\"Ui1\\" data-ui-rendered=\\"Ui1\\" class=\\"ecom-page-navigation pEn\\"><span><span class=\\"ecom-button-wrap btn u-display-hidden\\"><button class=\\"ecom-button\\"><span><span>CONTINUE</span>
  <ecomarrow-stub direction=\\"right\\" size=\\"lg\\" class=\\"arrow\\"></ecomarrow-stub>
  </span></button></span></span>
  <!---->
  <p class=\\"legal\\">
    NO PURCHASE OR FEE NECESSARY TO ENTER.
    <br>A PURCHASE WON’T IMPROVE AN INDIVIDUAL’S CHANCE OF WINNING.
  </p>
</div>"
`;

exports[`[Ui2] PageEndNavigation with 2 buttons matches snapshot 1`] = `
"<div data-ui-api=\\"Ui2\\" data-ui-rendered=\\"Ui2\\" class=\\"ecom-page-navigation pEn\\"><span><span class=\\"ecom-button-wrap btn u-display-hidden\\"><button class=\\"ecom-button\\"><span><span>CONTINUE SHOPPING</span>
  <ecomarrow-stub direction=\\"right\\" size=\\"lg\\" class=\\"arrow\\"></ecomarrow-stub>
  </span></button></span></span> <span><form method=\\"POST\\" novalidate=\\"novalidate\\" class=\\"g-w-100\\"><span class=\\"ecom-button-wrap btn u-display-hidden\\"><button class=\\"ecom-button\\"><span><span>ENTER NOW</span>
  <ecomarrow-stub direction=\\"right\\" size=\\"lg\\" class=\\"arrow\\"></ecomarrow-stub>
  </span></button></span></form></span>
  <p class=\\"legal\\">
    NO PURCHASE OR FEE NECESSARY TO ENTER.
    <br>A PURCHASE WON’T IMPROVE AN INDIVIDUAL’S CHANCE OF WINNING.
  </p>
</div>"
`;

exports[`[Ui3] PageEndNavigation with 1 button which will POST matches snapshot 1`] = `
"<div data-ui-api=\\"Ui3\\" data-ui-rendered=\\"Ui3\\" class=\\"ecom-page-navigation pEn\\"><span><form method=\\"POST\\" novalidate=\\"novalidate\\" class=\\"g-w-100\\"><span class=\\"ecom-button-wrap btn u-display-hidden\\"><button class=\\"ecom-button\\"><span><span>CONTINUE</span>
  <ecomarrow-stub direction=\\"right\\" size=\\"lg\\" class=\\"arrow\\"></ecomarrow-stub>
  </span></button></span></form></span>
  <!---->
  <p class=\\"legal\\">
    NO PURCHASE OR FEE NECESSARY TO ENTER.
    <br>A PURCHASE WON’T IMPROVE AN INDIVIDUAL’S CHANCE OF WINNING.
  </p>
</div>"
`;

exports[`PageEndNavigation matches snapshot 1`] = `
"<div class=\\"ecom-page-navigation pEn\\"><span><span class=\\"ecom-button-wrap btn u-display-hidden\\"><button class=\\"ecom-button\\"><span><span></span>
  <ecomarrow-stub direction=\\"right\\" size=\\"lg\\" class=\\"arrow\\"></ecomarrow-stub>
  </span></button></span></span>
  <!---->
  <p class=\\"legal\\">
    NO PURCHASE OR FEE NECESSARY TO ENTER.
    <br>A PURCHASE WON’T IMPROVE AN INDIVIDUAL’S CHANCE OF WINNING.
  </p>
</div>"
`;
